src/mpid/ch4/shm/ipc/gpu/gpu_post.c

```diff
@@ -172,6 +172,10 @@ int MPIDI_GPU_get_ipc_attr(const void *vaddr, int rank, MPIR_Comm * comm,
 
     ipc_attr->ipc_handle.gpu.global_dev_id = global_dev_id;
     ipc_attr->threshold.send_lmt_sz = MPIR_CVAR_CH4_IPC_GPU_P2P_THRESHOLD;
+
+    if (handle_obj == NULL) {
+        MPIDI_GPU_ipc_handle_cache_insert(rank, comm, ipc_attr->ipc_handle.gpu);
```

Comment:
Why not do this right after `MPL_gpu_ip_handle_create` after line 157?

Reply:
This was because at that point, ipc_attr hasn't been fully set up yet.

Reply:
Why don't we move line 143-161 after line 174?

Reply:
It might work, I will give it a try.

Reply:
I made the change

Reply:
The cache only depend on `dev_id`, and `remote_base_addr`, right? Can we amend `MPIDI_GPU_ipc_handle_cache_insert` so it explicitly has parameters of `dev_id` and `remote_base_addr`? I think that way the code logic is more explicit and also more flexible to move the code around.

Reply:
right, I am not very familiar with how the cache is implemented. I would prefer not doing big change to cache in this PR. 
One thing I would propose (which require some refactoring) is that this function is called to determine whether to do IPC or not, the only thing needed from this call for that purpose is ipc_type and threshold.send_lmt_sz in order to make the decision. However this function has a side effect of creating an IPC handle, which means for small messages that we don't do IPC, an IPC handle is always created which is unnecessary, and has the overhead added to very small message latency.

Reply:
Make sense. I will look into it sometime later (independent of the PR).